File: test-junit/src/test/java/io/micronaut/test/junit5/MathInnerService2Test.java

package io.micronaut.test.junit5;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.annotation.MockBean;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;

@MicronautTest
class MathInnerService2Test {

    @Inject
    MathService mathService;

    @Inject
    MathService[] services;

    /**
     * Tests that it is possible to have 2 mock beans
     */
    @Test
    void testInnerMockAgain() {
        final int result = mathService.compute(10);

        Assertions.assertEquals(
                1,
                services.length
        );
        Assertions.assertEquals(
                50,
                result
        );
        Assertions.assertTrue(mathService instanceof MyService);
    }

    @MockBean(MathServiceImpl.class)
    static class MyService implements MathService {

        @Override
        public Integer compute(Integer num) {
            return 50;
        }
    }
}

File: test-kotest/src/test/kotlin/io/micronaut/test/kotest/MathInnerService2Test.kt
package io.micronaut.test.kotest

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.micronaut.test.annotation.MicronautTest
import io.micronaut.test.annotation.MockBean

@MicronautTest
class MathInnerService2Test(
        private val mathService: MathService,
        private val services: Array<MathService>
): BehaviorSpec({

    given("an inner class mock") {

        `when`("the mock is called") {
            val result = mathService.compute(10)

            then("the mock is used") {
                result shouldBe 50
                mathService is MyService
                services.size shouldBe 1
            }

        }
    }

}) {

    @MockBean(MathServiceImpl::class)
    open class MyService : MathService {

        override fun compute(num: Int): Int {
            return num * 5
        }
    }
}

